### PR TITLE
feat: add date value to reference lines

### PR DESCRIPTION
**Pull request by rephus**
Closes: #3956

### Description:
- [x] timestamp or date fields are available to select as the field to base my reference line on. 
- [x] The value-picker looks like the value-picker for `is equal to` in date filters 

[Screencast from 12-01-23 15:47:28.webm](https://user-images.githubusercontent.com/1983672/212109496-a5df08bb-b3ee-4e9f-8d74-b26712c2a802.webm)

- [x] Week picker :

Screenshot from 2023-01-13 10-17-29
Screenshot from 2023-01-13 10-21-06

- [x] Month picker

Screenshot from 2023-01-13 09-46-48

- [x] Year picker 

Screenshot from 2023-01-13 10-08-56

I think it looks missplaced because of the thick bar

- [x] You can pick dates that are plotted outisde of the chart (picking a date plotted outside of a chart just forces the axes to have a new max/min value) - similar to this issue #4117 (depends on [feat-adjust-axis-limits-with-reference-lines](https://github.com/lightdash/lightdash/pull/4144)  )

### CODE REVIEW

**Review by TuringLovesDeathMetal — APPROVED** [review]
Nice, this looks great @rephus ! 😍 Also, comparing this UI for reference lines to the one I just reviewed in the older PR 😚 👌 _so_ much nicer!

File: packages/frontend/src/components/ChartConfigPanel/Legend/ReferenceLine.tsx [hunk 1]
```diff
@@ -1,26 +1,40 @@
...
+    TimeFrames,
+    WeekDay,
 } from '@lightdash/common';
 import debounce from 'lodash/debounce';
+import moment from 'moment';
 import { FC, useCallback, useMemo, useState } from 'react';
 import FieldAutoComplete from '../../common/Filters/FieldAutoComplete';
+import MonthAndYearInput from '../../common/MonthAndYearInput';
+import { ReferenceLineField } from '../../common/ReferenceLine';
 import { Flex } from '../../common/ResourceList/ResourceTable/ResourceTable.styles';
```
Comment (IrakliJani) [inline packages/frontend/src/components/ChartConfigPanel/Legend/ReferenceLine.tsx:25]: I think we should duplicate this or make this flex a common component which will be reused both in ResourceTable and ReferenceLine

**Review by IrakliJani — COMMENTED** [review]
code looks good. added one minor comment
I also found one UI bug and one feature bug (possibly?)

1. label is off the bounds
CleanShot 2023-01-17 at 16 17 07@2x

2. can't see a reference line and chart looks weird. am I doing something wrong?

https://user-images.githubusercontent.com/962095/212897818-5d1d5205-e5bf-4277-ac04-e3fc8df56a19.mp4

**Comment by rephus** [comment]
@IrakliJani  

1. This is a problem with echarts, echart doesn't handle things out of bounds very well. But we can still add a margin manually. So there is not much we can do (We have a few issues opened on their repo about this) 

2. Good catch. This was a merging issue from another branch where I automatically adjust the axis based on the min/max values. I fixed this and now it works with dates too !
Screenshot from 2023-01-17 16-06-44

**Review by IrakliJani — COMMENTED** [review]
- I think there is a bug with min/max counting. after some amount - in this case 9 - 9 becomes the max and trims the chart

https://user-images.githubusercontent.com/962095/213155421-30c9336e-f308-406d-a012-c80840a5f237.mp4

- above seems to be fine with flipped chart. altho modal does not persist the state (might have the same problem with conditional formatting). problem is that chart config overlaps with the chart and when I want to see the reference line I close the modal. reference line is visible but the value is gone. (more like an UX bug I guess)

File: packages/frontend/src/hooks/echarts/useEcharts.ts [hunk 2]
```diff
@@ -290,27 +297,66 @@ const getMinAndMaxReferenceLines = (
...
+                            try {
+                                const axisValue =
+                                    axis === 'yAxis' ? data.yAxis : data.xAxis;
+                                if (axisValue === undefined) return acc;
+                                const value = parseInt(axisValue);
+                                if (isNaN(value)) return acc;
+                                return [...acc, value];
+                            } catch (e) {
+                                return acc;
+                            }
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/frontend/src/hooks/echarts/useEcharts.ts:330]: Maybe if the data is undefined and cant' be parsed to a number ? I don't want the chart/page to break if we can't detect the min/max of a value...

**Comment by rephus** [comment]
> I think there is a bug with min/max counting. after some amount - in this case 9 - 9 becomes the max and trims the chart

Whaaaaaat, how crazy, let me take a look 


> altho modal does not persist the state 

There is a ticket for taht already, I'll take a look after this one is merged.

**Comment by rephus** [comment]
@IrakliJani I fixed this bug , there was a problem with float numbers. I added a frontend unit test for that.

**Review by IrakliJani — APPROVED** [review]
looks good! nice work 🙌 🙌 🙌

**Pull request merged**